Events api docs (#821)
boostercloud/booster: docs/chapters/04_features.md

> [!NOTE]  Remember that, in case you want to subscribe to a read model that is restricted to a specific set of roles, you must send the **access token** retrieved upon sign-in. Check ["Authorizing operations"](#authorizing-operations) to know how to do this.

### Reading events

You can also fetch events directly if you need. To do so, there are two kind of queries that have the following structure:
```graphql
query {
  eventsByEntity(entity: <name of entity>, entityID: "<id of the entity>") {
    selection_field_list
  }
}

query {
  eventsByType(type: <name of event>) {
    selection_field_list
  }
}
```
Where:

> - _<name of your entity>_ is the name of the class corresponding to the entity whose events you want to retrieve.

Reviewer: For some reason, all the words between < and > are not being rendered in the final ReadMe. I guess it is because they are interpreted as HTML tags and they are stripped. May be using the special symbols &gt and the like would work (check how it was done in the other places 😉 )

- _<id of the entity>_ is the ID of the specific entity instance whose events you are interested in. **This is optional**
- _<name of event>_ is the name of the class corresponding to the event type whose instances you want to retrieve.
- _selection_field_list_ is a list with the names of the specific fields you want to get as response. See the response example below to know more.

#### Examples

```
  URL: "<graphqlURL>"
```
**A) Read all events associated with CartEntity by a specific ID**
```graphql
query {

>   eventsByEntity(entity: CartEntity, entityID: "B5") {

Reviewer: ```suggestion
  eventsByEntity(entity: Cart, entityID: "ABC123") {
```

    type entity entityID requestID createdAt value
  }
}
```

> **B) Read all events from CartEntity**

Reviewer: For consistency, we should write this in the same way as case A but leaving clear this time is without the ID part. If not, it seems it is something completely different:
```suggestion
**B) Read all events associated with any instance of the entity Cart**
```

```graphql
query {

>   eventsByEntity(entity: CartEntity) {

Reviewer: ```suggestion
  eventsByEntity(entity: Cart) {
```

    type entity entityID requestID createdAt value
  }
}
```

> For cases A and B, the response look like this:

Reviewer: So far, we have only seen cases A and B, so we don't need to mention them. Here is my suggestion with more description about the response:

```suggestion
For these cases, you would get an array of event _envelopes_ as a response. This means that you get some metadata related to the event along with the event content, which can be found inside the `"value"` field.
The response would look like this:
```

```json
{
    "data": {
        "eventsByEntity": [
            {
                "type": "CartItemChanged",
                "entity": "Cart",

>                 "entityID": "B5",

Reviewer: ```suggestion
                "entityID": "ABC123",
```
Nitpicking: just to make it look more like an ID. Feel free to take it or leave it

                "requestID": "7a9cc6a7-7c7f-4ef0-aef1-b226ae4d94fa",
                "createdAt": "2021-05-12T08:41:13.792Z",
                "value": {
                    "productId": "73f7818c-f83e-4482-be49-339c004b6fdf",

>                     "cartId": "B5",

Reviewer: ```suggestion
                    "cartId": "ABC123",
```

                    "quantity": 2
                }
            }
        ]
    }
}
```

> **C) Query specific events, no matter the entity/es it has assigned.**

Reviewer: This is not completely correct, because you can't query a specific event (an instance), but a specific event _type_. Maybe it is a bit nitpicky, but it is better to be nitpicking in the docs to avoid confusion.

Also, I suggest to use the same style as other examples, where you used the word "Read" instead of "Query":

```suggestion
**C) Read events of a specific type**
```

```graphql
query {

>   eventsByType(type: CartChangedEvent) {

Reviewer: To be consistent with the response and the previous examples.
```suggestion
  eventsByType(type: CartItemChanged) {
```

    type entity entityID requestID createdAt value
  }
}
```

> In this case, for the **eventsByType** endpoint, the response would be:
> ```json
> {
>     "data": {
>         "eventsByType": [
>             {
>                 "type": "CartItemChanged",
>                 "entity": "Cart",
>                 "entityID": "5d81a60c-2e27-4fef-9f57-e5ae36519144",
>                 "requestID": "7a9cc6a7-7c7f-4ef0-aef1-b226ae4d94fa",
>                 "createdAt": "2021-05-12T08:41:13.792Z",
>                 "value": {
>                     "productId": "73f7818c-f83e-4482-be49-339c004b6fdf",
>                     "cartId": "5d81a60c-2e27-4fef-9f57-e5ae36519144",
>                     "quantity": 2
>                 }
>             }, 
>           ...
>         ]}
>     }
> }
> ```

Reviewer: As the response is exactly the same as before, I would not include it this time, but would write some explanation
```suggestion
The response would have the same structure as seen in the previous examples. The only difference is that this time you will get only the events with the type you have specified ("CartItemChanged")
```

#### Time filters

> Optionally, you can get events in a range of dates for any endpoint of this API. The time filter format is in ISO format, with any precision, for example:

Reviewer: Remember that in GraphQL, we talk about mutations or queries, not endpoints, as we have only one endpoint for everything.
```suggestion
Optionally, for any of the previous queries, you can include a `from` and/or `to` time filters to get only those events that happened inside that time range. You must use a string with a time in ISO format with any precision you like, for example:
```

> * from:"2021" : Events created on 2021 year or up

Reviewer: ```suggestion
* `from:"2021"` : Events created on 2021 year or up
```

> * from:"2021-02-12" to:"2021-02-13" : Events created during February 12th

Reviewer: ```suggestion
* `from:"2021-02-12" to:"2021-02-13"` : Events created during February 12th
```

> * from:"2021-03-16T16:16:25.178" : Events created at that date and time, using millisecond precision

Reviewer: ```suggestion
* `from:"2021-03-16T16:16:25.178"`: Events created at that date and time, using millisecond precision, or later.
```

#### Time filters examples

> **A) CartEntity events from February 23rd to July 20th, 2021**

Reviewer: ```suggestion
**A) Cart events from February 23rd to July 20th, 2021**
```

```graphql
query {

>   eventsByEntity(entity: CartEntity, from:"2021-02-23", to:"2021-07-20") {

Reviewer: ```suggestion
  eventsByEntity(entity: Cart, from:"2021-02-23", to:"2021-07-20") {
```

For a force delete without asking for confirmation, you can run `boost nuke -e <environment name> -f`.

> [!ATTENTION]  Be EXTRA CAUTIOUS with this option, all your application data will be irreversibly DELETED without confirmation.

## Events API

Booster now supports fetching events from your Booster application! Here are some examples of how it works:

> 

Author: I didn't put the HTTP request because we have more examples in other places and it looks pretty much the same. Also changed a bit the structure, let me know what you think @alvaroloes 🙏
